src/hotspot/share/gc/g1/g1RootProcessor.cpp

```diff
@@ -59,6 +59,12 @@ void G1RootProcessor::evacuate_roots(G1ParScanThreadState* pss, uint worker_id)
   G1EvacuationRootClosures* closures = pss->closures();
   process_java_roots(closures, phase_times, worker_id);
 
+#ifdef ASSERT
+  // only for verification purpose
+  // already processed in java roots.
+  _process_strong_tasks.try_claim_task(G1RP_PS_CodeCache_oops_do);
+#endif
```

Comment:
Rather than these fake claims, consider something like this:

```
template<typename... Ts>
void all_tasks_completed(uint nworkers, Ts... tags) {
  // Type-check more_skipped are all of the same type as first_skipped.
  T0 typed_skipped[] = { first_skipped, more_skipped... };
  uint skipped[] = { static_cast<uint>(tags)... };
  all_tasks_completed_impl(nworkers, skipped, ARRAY_SIZE(skipped));
}

void all_tasks_completed(uint nworkers) {
  all_tasks_completed_impl(nworkers, nullptr, 0);
}
```

Usage:

```
all_tasks_completed(n_workers(),
                    G1RP_PS_CodeCache_oops_do,
                    G1RP_PS_refProcessor_oops_do)
```

all_tasks_completed_impl can check that all tasks have been claimed except
the skipped ones, which have not been claimed.

There might be better ways to write the variadic all_tasks_completed.  It's
been a while since I've done anything with variadic templates.

Reply:
Here's a better version of the variadic `all_tasks_completed`

```
template<typename T0, typename... Ts,
         ENABLE_IF(Conjunction<std::is_same<T0, Ts>...>::value)>
void all_tasks_completed(uint n_threads, T0 first_skipped, Ts... more_skipped) {
  static_assert(std::is_convertible<T0, uint>::value, "not convertible");
  uint skipped[] = { static_cast<uint>(first_skipped), static_cast<uint>(more_skipped)... };
  all_tasks_completed_impl(n_threads, skipped, ARRAY_SIZE(skipped));
}
```
`Conjunction` is in metaprogramming/logical.hpp.

Reply:
Thank you; updated as suggested.